### PR TITLE
feat(plasma-b2c, plasma-web): focusTrap hook added

### PR DESCRIPTION
**Pull request by kayman233**
Добавлен хук useFocusTrap для захвата фокуса внутри одного DOM элемента

``` js
const ref = useFocusTrap();
return (
  <div ref={ref}>
    <button>Trapped button</button>
  </div>
)

```

Принимает параметры `active`(по умолчанию `true`) и `firstFocusSelector?: string | HTMLElement` - selector или Element, который должен быть в фокусе при открытии.

Также поправлено поведение в ci для юнит тестов в react17. При bootstrap теперь добавлены `plasma-hope`, `plasma-core`. Проблема заключалась в том, что без этого в зависимость подтягивались версии этих пакетом из npm, из-за чего возникали ошибки при билде.
 

:baby_chick: Download canary assets:

[color_asdk_ios-swift--canary.551.5331754583.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_asdk_ios-swift--canary.551.5331754583.swift)  
[color_asdk_kotlin--canary.551.5331754583.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_asdk_kotlin--canary.551.5331754583.kt)  
[color_asdk_react-native--canary.551.5331754583.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_asdk_react-native--canary.551.5331754583.ts)  
[color_sberHealth_ios-swift--canary.551.5331754583.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberHealth_ios-swift--canary.551.5331754583.swift)  
[color_sberHealth_kotlin--canary.551.5331754583.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberHealth_kotlin--canary.551.5331754583.kt)  
[color_sberHealth_react-native--canary.551.5331754583.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberHealth_react-native--canary.551.5331754583.ts)  
[color_sbermarket_business_ios-swift--canary.551.5331754583.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_business_ios-swift--canary.551.5331754583.swift)  
[color_sbermarket_business_kotlin--canary.551.5331754583.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_business_kotlin--canary.551.5331754583.kt)  
[color_sbermarket_business_react-native--canary.551.5331754583.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_business_react-native--canary.551.5331754583.ts)  
[color_sbermarket_ios-swift--canary.551.5331754583.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_ios-swift--canary.551.5331754583.swift)  
[color_sbermarket_kotlin--canary.551.5331754583.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_kotlin--canary.551.5331754583.kt)  
[color_sbermarket_metro_ios-swift--canary.551.5331754583.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_metro_ios-swift--canary.551.5331754583.swift)  
[color_sbermarket_metro_kotlin--canary.551.5331754583.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_metro_kotlin--canary.551.5331754583.kt)  
[color_sbermarket_metro_react-native--canary.551.5331754583.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_metro_react-native--canary.551.5331754583.ts)  
[color_sbermarket_react-native--canary.551.5331754583.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_react-native--canary.551.5331754583.ts)  
[color_sbermarket_selgros_ios-swift--canary.551.5331754583.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_selgros_ios-swift--canary.551.5331754583.swift)  
[color_sbermarket_selgros_kotlin--canary.551.5331754583.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_selgros_kotlin--canary.551.5331754583.kt)  
[color_sbermarket_selgros_react-native--canary.551.5331754583.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_selgros_react-native--canary.551.5331754583.ts)  
[color_sbermarket_wlbusiness_ios-swift--canary.551.5331754583.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_wlbusiness_ios-swift--canary.551.5331754583.swift)  
[color_sbermarket_wlbusiness_kotlin--canary.551.5331754583.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_wlbusiness_kotlin--canary.551.5331754583.kt)  
[color_sbermarket_wlbusiness_react-native--canary.551.5331754583.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_wlbusiness_react-native--canary.551.5331754583.ts)  
[color_sberonline_ios-swift--canary.551.5331754583.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberonline_ios-swift--canary.551.5331754583.swift)  
[color_sberonline_kotlin--canary.551.5331754583.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberonline_kotlin--canary.551.5331754583.kt)  
[color_sberonline_react-native--canary.551.5331754583.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberonline_react-native--canary.551.5331754583.ts)  
[color_sberprime_ios-swift--canary.551.5331754583.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_ios-swift--canary.551.5331754583.swift)  
[color_sberprime_kotlin--canary.551.5331754583.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_kotlin--canary.551.5331754583.kt)  
[color_sberprime_react-native--canary.551.5331754583.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_react-native--canary.551.5331754583.ts)  
[shadow_sbermarket_react-native--canary.551.5331754583.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/shadow_sbermarket_react-native--canary.551.5331754583.ts)  
[typo_mage_ios-swift--canary.551.5331754583.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_ios-swift--canary.551.5331754583.swift)  
[typo_mage_kotlin--canary.551.5331754583.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_kotlin--canary.551.5331754583.kt)  
[typo_mage_react-native--canary.551.5331754583.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_react-native--canary.551.5331754583.ts)  
[typo_plasma_ios-swift--canary.551.5331754583.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_ios-swift--canary.551.5331754583.swift)  
[typo_plasma_kotlin--canary.551.5331754583.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_kotlin--canary.551.5331754583.kt)  
[typo_plasma_react-native--canary.551.5331754583.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_react-native--canary.551.5331754583.ts)  
[typo_ruler_ios-swift--canary.551.5331754583.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_ios-swift--canary.551.5331754583.swift)  
[typo_ruler_kotlin--canary.551.5331754583.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_kotlin--canary.551.5331754583.kt)  
[typo_ruler_react-native--canary.551.5331754583.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_react-native--canary.551.5331754583.ts)  
[typo_sage_ios-swift--canary.551.5331754583.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_ios-swift--canary.551.5331754583.swift)  
[typo_sage_kotlin--canary.551.5331754583.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_kotlin--canary.551.5331754583.kt)  
[typo_sage_react-native--canary.551.5331754583.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_react-native--canary.551.5331754583.ts)  
[typo_sbermarket_ios-swift--canary.551.5331754583.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_ios-swift--canary.551.5331754583.swift)  
[typo_sbermarket_kotlin--canary.551.5331754583.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_kotlin--canary.551.5331754583.kt)  
[typo_sbermarket_react-native--canary.551.5331754583.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_react-native--canary.551.5331754583.ts)  
[typo_soulmate_ios-swift--canary.551.5331754583.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_ios-swift--canary.551.5331754583.swift)  
[typo_soulmate_kotlin--canary.551.5331754583.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_kotlin--canary.551.5331754583.kt)  
[typo_soulmate_react-native--canary.551.5331754583.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_react-native--canary.551.5331754583.ts)

<details>
  <summary>📦 Published PR as canary version: <code>Canary Versions</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @salutejs/plasma-b2c@1.217.1-canary.551.5331754583.0
  npm install @salutejs/plasma-hope@1.217.1-canary.551.5331754583.0
  npm install @salutejs/plasma-web@1.217.1-canary.551.5331754583.0
  # or 
  yarn add @salutejs/plasma-b2c@1.217.1-canary.551.5331754583.0
  yarn add @salutejs/plasma-hope@1.217.1-canary.551.5331754583.0
  yarn add @salutejs/plasma-web@1.217.1-canary.551.5331754583.0
  ```
</details>

### CODE REVIEW

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-551/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-551/

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-551/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-551/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-551/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-551/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-551/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-551/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-551/temple-storybook/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-551/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-551/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-551/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-551/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-551/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-551/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-551/temple-storybook/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-551/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-551/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-551/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-551/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-551/temple-storybook/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-551/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-551/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-551/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-551/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-551/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-551/temple-storybook/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-551/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-551/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-551/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-551/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-551/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-551/temple-storybook/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

Top 5 significant render changes:
packages/plasma-web/src/components/Modal/Modal.perftest.tsx#Default: -22.40%

File: packages/plasma-hope/src/components/Modal/ModalView.tsx [hunk 1]
```diff
@@ -1,6 +1,5 @@
 import React from 'react';
 import styled from 'styled-components';
-import FocusLock from 'react-focus-lock';
```
Comment (Yeti-or) [inline packages/plasma-hope/src/components/Modal/ModalView.tsx:3]: значит из зависимостей это можно удалить?

File: packages/plasma-hope/src/utils/ariaHider.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
+type Value =
+    | {
+          node: HTMLElement;
+          ariaHidden: string | null;
+      }
+    | undefined;
+
+/**
+ * Установка `aria-hidden` на всех элементах, кроме активной ноды.
```
Comment (Yeti-or) [inline packages/plasma-hope/src/utils/ariaHider.ts:9]: насколько я понимаю тут на всех => на всех непосредственных детях body?

а если поставить сразу на body?

File: packages/plasma-web/src/components/Modal/Modal.component-test.tsx [hunk 3]
```diff
@@ -166,4 +166,35 @@ describe('plasma-web: Modal', () => {
 
         cy.matchImageSnapshot();
     });
+
+    it('check focus trap', () => {
```
Comment (Yeti-or) [inline packages/plasma-web/src/components/Modal/Modal.component-test.tsx:170]: А тесты на сам focus trap?

Если передали focusable элемент,
если не нашли tabbale потомком

если нашли focusable
или focusable сам элемент?

File: packages/plasma-hope/src/hooks/useFocusTrap.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                    if (node.ownerDocument) {
+                        processNode(node);
+                    }
+                });
+
+                ref.current = node;
+            } else {
+                ref.current = null;
+            }
+        },
```
Comment (neretin-trike) [inline packages/plasma-hope/src/hooks/useFocusTrap.ts:79]: метод выглядит сильно перегруженным. Можешь пожалуйста попробовать избавиться от вложеннойстей if'ов и декомпозировать его на составные?

File: packages/plasma-web/src/components/Modal/Modal.component-test.tsx [hunk 5]
```diff
@@ -166,4 +166,35 @@ describe('plasma-web: Modal', () => {
...
+        cy.get('button').contains('Open modal B').type('{enter}');
+        cy.get('button').contains('Close-B').type('{enter}');
+        cy.focused().should(($p) => {
+            expect($p).to.contain('Open modal B');
+        });
+
+        cy.focused().tab();
+        cy.focused().tab();
+        // по какой-то причине таб не срабатывает с первого раза
+        cy.focused().tab().tab();
```
Comment (neretin-trike) [inline packages/plasma-web/src/components/Modal/Modal.component-test.tsx:190]: возможно, он при нажатии на tab переходит всё-таки на какой-то другой элемент

File: packages/plasma-hope/src/utils/focusManager.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+const focusLaterElements: Array<HTMLElement> = [];
+let focusElement: HTMLElement | null = null;
+let needToFocus = false;
+
+const handleBlur = () => {
+    needToFocus = true;
+};
+
+const handleFocus = () => {
+    if (needToFocus) {
```
Comment (neretin-trike) [inline packages/plasma-hope/src/utils/focusManager.ts:12]: давай тут инверсию сделаем

File: packages/plasma-hope/src/utils/scopeTab.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+ */
+export const scopeTab = (node: HTMLElement, event: KeyboardEvent) => {
+    const tabbable = findTabbableDescendants(node);
+    if (!tabbable.length) {
+        event.preventDefault();
+        return;
+    }
+
+    const finalTabbable = tabbable[event.shiftKey ? 0 : tabbable.length - 1];
+    const leavingFinalTabbable = finalTabbable === document.activeElement || node === document.activeElement;
```
[the remark this answers is not in the document]
Comment (kayman233) [reply packages/plasma-hope/src/utils/scopeTab.ts:16]: вообще у меня такой кейс не воспроизводился, но утверждается что это нужно для обработки случая, когда нажимается shift+tab после открытия с помощью мышки

File: cypress/support/index.ts [hunk 8]
```diff
@@ -3,6 +3,8 @@ import '@cypress/code-coverage/support';
 // по какой-то причине axe подключается только через require
 // иначе не поднимается dev-server webpack'а
 require('cypress-axe');
+// для эмуляции таба
+require('cypress-plugin-tab');
```
Comment (Yakutoc) [inline cypress/support/index.ts:7]: [Уточнение]

То есть нажатие клавиши Tab никак по другому не обработать?

File: packages/plasma-hope/src/hooks/useFocusTrap.ts [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+                        focusElement =
+                            typeof options.focusSelector === 'string'
+                                ? node.querySelector(options.focusSelector)
+                                : options.focusSelector;
+                    }
+
+                    if (!focusElement) {
+                        const children = Array.from<HTMLElement>(node.querySelectorAll(focusSelector));
+                        focusElement =
+                            // tabbable предпочтительнее focusable
```
Comment (Yakutoc) [inline packages/plasma-hope/src/hooks/useFocusTrap.ts:53]: [Уточнение]

А почему предпочтительнее?

File: packages/plasma-hope/src/hooks/useFocusTrap.ts [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        [active, options.focusSelector, options.disableAriaHider],
+    );
+
+    useEffect(() => {
+        if (!active) {
+            return;
+        }
+
+        const handleKeyDown = (event: KeyboardEvent) => {
+            if (event.key === 'Tab' && ref.current) {
```
Comment (Yakutoc) [inline packages/plasma-hope/src/hooks/useFocusTrap.ts:89]: [Уточнение]

`event.key === 'Tab'` <--- Во всех браузерах результат будет один и тот же? Нигде нет разночтения tab, Tab, etc

File: packages/plasma-hope/src/components/Modal/ModalView.tsx (hunk 1 above)
Comment (kayman233) [reply packages/plasma-hope/src/components/Modal/ModalView.tsx:3]: убрал из зависимостей

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-551/

File: cypress/support/index.ts (hunk 8 above)
Comment (kayman233) [reply cypress/support/index.ts:7]: есть стандартные триггеры в сайпресе, но они почему-то у меня совсем не хотели работать
https://github.com/cypress-io/cypress/issues/299
Вообще, эта боль с tab-ами уже давно. Я перепробовал другие плагины, этот оказался наиболее рабочим (хотя тоже видимо с косяками, 2 комментами выше с Артемом как раз нашли)

File: packages/plasma-hope/src/hooks/useFocusTrap.ts (hunk 9 above)
Comment (kayman233) [reply packages/plasma-hope/src/hooks/useFocusTrap.ts:53]: ну у нас чисто в реализации isTabble это также и isFocusable. Отличие, что у первого еще есть tabIndex неотрицательный

File: packages/plasma-hope/src/hooks/useFocusTrap.ts (hunk 10 above)
Comment (kayman233) [reply packages/plasma-hope/src/hooks/useFocusTrap.ts:89]: https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key
кажется во всех, да keyCode уже deprecated

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-551/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-551/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-551/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-551/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-551/temple-storybook/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

File: packages/plasma-hope/src/hooks/useFocusTrap.ts (hunk 4 above)
Comment (kayman233) [reply packages/plasma-hope/src/hooks/useFocusTrap.ts:64]: вроде немного поменьше стал, кажется без if'ов тут никуда

File: packages/plasma-web/src/components/Modal/Modal.component-test.tsx (hunk 3 above)
Comment (kayman233) [reply packages/plasma-web/src/components/Modal/Modal.component-test.tsx:170]: добавил отдельные тесты для хука

File: packages/plasma-hope/src/utils/tabbable.ts [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+].join(',');
+
+const hidden = (el: HTMLElement) => {
+    return (el.offsetWidth <= 0 && el.offsetHeight <= 0) || el.style.display === 'none';
+};
+
+export const visible = (element: HTMLElement) => {
+    let parentElement: HTMLElement = element;
+
+    while (parentElement) {
```
[the remark this answers is not in the document]
Comment (kayman233) [reply packages/plasma-hope/src/utils/tabbable.ts:28]: в итоге делаю проверку до контейнера, а не до body

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-551/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-551/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-551/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-551/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-551/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-551/temple-storybook/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

**Review by Yakutoc — APPROVED** [review]

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-551/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-551/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-551/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-551/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-551/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-551/temple-storybook/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

File: website/plasma-web-docs/docs/functions/hooks.mdx [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+```
+
+Принимает параметры `active`(по умолчанию `true`) и `firstFocusSelector?: string | HTMLElement` - selector или Element, который должен быть в фокусе при открытии.
+
+```tsx live
+import React, { useState } from "react";
+import { useFocusTrap, Button } from "@salutejs/plasma-web";
+
+export function App() {
+    const ref = useFocusTrap(true, { focusSelector: "#focus-first" });
```
Comment (neretin-trike) [inline website/plasma-web-docs/docs/functions/hooks.mdx:41]: тут видимо должен называться `firstFocusSelector`?

Comment (neretin-trike) [reply website/plasma-web-docs/docs/functions/hooks.mdx:41]: и передаваться просто строкой, а не объектом

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-551/

File: packages/plasma-hope/src/hooks/useFocusTrap.ts (hunk 4 above)
Comment (neretin-trike) [reply packages/plasma-hope/src/hooks/useFocusTrap.ts:64]: в целом да, стало лучше. Но давай ещё вытащим метод processNode наружу (кажется, что это можно сделать), т.е. просто перенесём из setRef, и в конце после `ref.current = node;` предлагаю сделать return, чтобы избавиться от else

File: packages/plasma-hope/src/hooks/useFocusTrap.ts [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+                teardownScopedFocus();
+                returnFocus();
+            }
+
+            if (active && node) {
+                setupScopedFocus(node);
+                markForFocusLater();
+
+                // Находим первый элемент для фокуса
+                const processNode = (node: HTMLElement) => {
```
Comment (neretin-trike) [inline packages/plasma-hope/src/hooks/useFocusTrap.ts:28]: вообще с точки зрения назначения самого метода кажется, что можно выделить чисто поиск элемента, а фокус - отдельно в случае успеха. И получится что-то вроде

```ts
getFocusElement = () => { ... весь твой код ниже до коммента 'Фокус на элемент внутри' }
const focusElement = getFocusElement();
if (focusElement) {
  focusElement.focus();
}
```

File: packages/plasma-hope/src/utils/focusManager.ts [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    focusElement = element;
+    document.addEventListener('focusout', handleBlur, false);
+    document.addEventListener('focusin', handleFocus, true);
+};
+
+export const teardownScopedFocus = () => {
+    focusElement = null;
+    document.removeEventListener('focusout', handleBlur);
+    document.removeEventListener('focusin', handleFocus);
+};
```
Comment (neretin-trike) [inline packages/plasma-hope/src/utils/focusManager.ts:46]: выглядит как файл, который должен быть классом, а не просто набором методов с глобальными переменными 🤔 Что думаешь по этому поводу?

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-551/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-551/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-551/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-551/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-551/temple-storybook/

File: packages/plasma-hope/src/hooks/useFocusTrap.ts [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+
+import { markForFocusLater, returnFocus, setupScopedFocus, teardownScopedFocus } from '../utils/focusManager';
+import { focusSelector, isFocusable, isTabble } from '../utils/tabbable';
+import { scopeTab } from '../utils/scopeTab';
+
+/**
+ *  Захватывает фокус внутри DOM node.
+ * */
+export const useFocusTrap = (
+    active = true,
```
Comment (Yeti-or) [inline packages/plasma-hope/src/hooks/useFocusTrap.ts:11]: А почему мне может хотеться переключать active между true и false ?

Comment (Yeti-or) [reply packages/plasma-hope/src/hooks/useFocusTrap.ts:11]: и в целом useFocusTrap(false) это в каком случае использовать?

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

File: packages/plasma-hope/src/utils/focusManager.ts [hunk 16]
```diff
@@ -0,0 +1,46 @@
+import { findTabbableDescendants } from './tabbable';
+
```
Comment (Yeti-or) [inline packages/plasma-hope/src/utils/focusManager.ts:2]: все же здесь не хватает комментов, что именно тут происходит и зачем; я помню что мы это обсуждали на встрече, но вот прошли выходные и я опять забыл всё =/

File: packages/plasma-web/src/hooks/useFocusTrap.component-test.tsx [hunk 17]
```diff
@@ -0,0 +1,109 @@
+import React, { useState } from 'react';
+import styled, { createGlobalStyle } from 'styled-components';
```
Comment (Yeti-or) [inline packages/plasma-web/src/hooks/useFocusTrap.component-test.tsx:2]: createGlobalStyle ?

File: website/plasma-web-docs/docs/functions/hooks.mdx [hunk 18]
```diff
@@ -0,0 +1,62 @@
+---
+id: hooks
+title: Хуки
+---
+
+# Хуки
+
+## `useFocusTrap`
```
Comment (Yeti-or) [inline website/plasma-web-docs/docs/functions/hooks.mdx:8]: надо бы тут отметить что он уже подключен к Modal

File: packages/plasma-hope/src/hooks/useFocusTrap.ts (hunk 15 above)
Comment (kayman233) [reply packages/plasma-hope/src/hooks/useFocusTrap.ts:11]: ну как мне показалось, такое может понадобиться, если кому-то нужно его выключить принудительно по флагу какому-то. В целом такой механизм во всех других реализациях тоже есть

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-551/

File: packages/plasma-hope/src/utils/focusManager.ts (hunk 16 above)
Comment (kayman233) [reply packages/plasma-hope/src/utils/focusManager.ts:2]: добавил комменты. Пока их писал, понял, что нашел баг. Он заключался в том, что когда закрываем вложенную trap ноду, то у исходная просто терялась, и при блюре(когда просто фокус убираем, мышкой например), то он фокус происходил уже по всему документу. Теперь немного изменил. У нас 2 стека - один отвечает за элементы, которые хотим зафокусить после анмаунта, а второй уже за эти самые ноды, в которых всегда должен находится фокус

Написал для этого тесты еще, чтобы зафиксировать

File: website/plasma-web-docs/docs/functions/hooks.mdx (hunk 18 above)
Comment (kayman233) [reply website/plasma-web-docs/docs/functions/hooks.mdx:8]: отметил

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-551/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-551/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-551/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-551/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-551/temple-storybook/

File: packages/plasma-hope/src/utils/focusManager.ts [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+/**
+ *  Менеджер фокуса при открытии и закрытии нод при использовании focus-trap.
+ *  Также необходим, чтобы фокус всегда должен находиться внутри необходимой ноды.
+ * */
+export class FocusManager {
+    // массив с элементами, которые нужно зафукосить после анмаунта
```
Comment (neretin-trike) [inline packages/plasma-hope/src/utils/focusManager.ts:8]: `зафукосить` => `зафокусить`

**Review by neretin-trike — APPROVED** [review]

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

**Review by Yeti-or — APPROVED** [review]

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-551/

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-551/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-551/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-551/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-551/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-551/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-551/temple-storybook/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

**Pull request merged**